File: app/src/main/java/com/example/features/main/MainActivity.java
package com.example.features.main;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.DrawableRes;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.widget.Toast;

import com.example.R;
import com.example.databinding.MainActivityBinding;
import com.example.features.BaseActivity;
import com.example.features.authentication.LoginActivity;
import com.example.features.profile.ProfileSummaryFragment;
import com.example.utils.AppUtils;
import com.newrelic.agent.android.NewRelic;

import java.util.Vector;

public class MainActivity extends BaseActivity implements TabLayout.OnTabSelectedListener, MainView {
	static final String CURRENT_TAB = "com.example.features.MainActivity.CURRENT_TAB";
	static final int REQUEST_CODE_RECOVER_PLAY_SERVICES = 1;

	MainActivityBinding mBinding;
	MainPresenter mPresenter;
	TabbedPagerAdapter mTabbedPagerAdapter;
	int mCurrentTab = 0;
	boolean mDoubleBackToExitPressedOnce = false;
	Handler mDoubleBackHandler = new Handler();
	Runnable mExitRunnable = (() -> mDoubleBackToExitPressedOnce = false);
	Toast mExitToast;

	public static Intent newIntent(Context context) {
		Intent intent = new Intent(context, MainActivity.class);
		intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);

		return intent;
	}

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		mPresenter = new MainPresenter(this);
		NewRelic.withApplicationToken(getString(R.string.app_newrelic_token)).start(this.getApplication());

		if (AppUtils.checkPlayServices(this, REQUEST_CODE_RECOVER_PLAY_SERVICES)) {
			mPresenter.checkAuthentication();
		}
	}

	@Override
	public void onBackPressed() {
		if (mDoubleBackToExitPressedOnce) {
			super.onBackPressed();

			if (mExitToast != null) {
				mExitToast.cancel();
			}

			finish();
			return;
		}

		mDoubleBackToExitPressedOnce = true;
		mExitToast = Toast.makeText(this, R.string.common_double_back_to_exit, Toast.LENGTH_SHORT);

		mExitToast.show();
		mDoubleBackHandler.postDelayed(mExitRunnable, 2000);
	}


	@Override
	protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putInt(CURRENT_TAB, mCurrentTab);
	}

	@Override
	protected void onRestoreInstanceState(Bundle savedInstanceState) {
		super.onRestoreInstanceState(savedInstanceState);
		mCurrentTab = savedInstanceState == null ? 0 : savedInstanceState.getInt(CURRENT_TAB);
	}

	@Override
	protected void onDestroy() {
		mDoubleBackHandler.removeCallbacks(mExitRunnable);
		super.onDestroy();
	}

	// MainView implementations

	@Override
	public void onAuthCheckResult(boolean isAuth) {
		if (!isAuth) {
			startActivity(LoginActivity.startLogin(this));
			finish();
			return;
		}

		setLayout();
	}

	// TabLayout event handlers

	@Override
	public void onTabSelected(TabLayout.Tab tab) {
		mBinding.tabPager.setCurrentItem(tab.getPosition());
		mCurrentTab = tab.getPosition();

		setTabColor(tab, ContextCompat.getColor(this, R.color.tab_seletected));
	}

	@Override
	public void onTabUnselected(TabLayout.Tab tab) {
		setTabColor(tab, ContextCompat.getColor(this, R.color.tab_seletected));
	}

	@Override
	public void onTabReselected(TabLayout.Tab tab) {

	}

	// Helper methods

	private void setTabIcon(TabLayout.Tab tab, @DrawableRes int icon) {
		if (tab == null) {
			return;
		}

		tab.setIcon(icon);
	}

	private void setTabColor(TabLayout.Tab tab, int color) {
		if (tab == null || tab.getIcon() == null) {
			return;
		}

		tab.getIcon().setColorFilter(color, PorterDuff.Mode.SRC_IN);
	}

	private void setLayout() {
		setTheme(R.style.AppTheme);
		mBinding = DataBindingUtil.setContentView(this, R.layout.main_activity);

		Vector<Fragment> fragments = new Vector<>();

		fragments.add(Fragment.instantiate(this, ProfileSummaryFragment.class.getName()));

		mTabbedPagerAdapter = new TabbedPagerAdapter(getSupportFragmentManager(), fragments);

		mBinding.tabPager.setAdapter(mTabbedPagerAdapter);
		mBinding.tabPager.setOffscreenPageLimit(1);

		mBinding.tablayoutMain.setupWithViewPager(mBinding.tabPager);
		mBinding.tablayoutMain.setTabGravity(TabLayout.GRAVITY_FILL);
		mBinding.tablayoutMain.addOnTabSelectedListener(this);

		mBinding.tabPager.setCurrentItem(mCurrentTab);

		setTabIcon(mBinding.tablayoutMain.getTabAt(0), R.drawable.ic_person);

		setTabColor(mBinding.tablayoutMain.getTabAt(0), ContextCompat.getColor(this, R.color.tab_unselected));

		setTabColor(mBinding.tablayoutMain.getTabAt(mCurrentTab), ContextCompat.getColor(this, R.color.tab_seletected));
	}
}

File: app/src/main/kotlin/br/com/vitorsalgado/example/features/main/MainActivity.kt
package br.com.vitorsalgado.example.features.main

import android.content.Context
import android.content.Intent
import android.graphics.PorterDuff
import android.os.Bundle
import android.os.Handler
import android.widget.Toast
import androidx.annotation.DrawableRes
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.findNavController
import br.com.vitorsalgado.example.R
import br.com.vitorsalgado.example.utils.AppUtils
import com.google.android.material.tabs.TabLayout

class MainActivity : AppCompatActivity() {
  private var mCurrentTab = 0
  private var mDoubleBackToExitPressedOnce = false
  private val mDoubleBackHandler = Handler()
  private val mExitRunnable = { mDoubleBackToExitPressedOnce = false }
  private var mExitToast: Toast? = null

  companion object {
    private const val REQUEST_CODE_RECOVER_PLAY_SERVICES = 9001
    private const val REQUEST_CODE_RECOVER_PLAY_SERVICES_WITHOUT_CALLBACK = 9002
    private const val REQUEST_CODE_WRITE_STORAGE_PERMISSION = 9003

    fun newIntent(context: Context): Intent {
      val intent = Intent(context, MainActivity::class.java)
      intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK

      return intent
    }
  }

  // region Activity Events

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    if (!AppUtils.checkPlayServices(this, REQUEST_CODE_RECOVER_PLAY_SERVICES)) {
      finish()
      return
    }

    setLayout()
  }

  override fun onSupportNavigateUp() = findNavController(R.id.navHost).navigateUp()

  override fun onBackPressed() {
    if (mDoubleBackToExitPressedOnce) {
      super.onBackPressed()

      if (mExitToast != null) {
        mExitToast!!.cancel()
      }

      finish()
      return
    }

    mDoubleBackToExitPressedOnce = true
    mExitToast = Toast.makeText(this, R.string.double_back_to_exit, Toast.LENGTH_SHORT)

    mExitToast!!.show()
    mDoubleBackHandler.postDelayed(mExitRunnable, 2000)
  }

  override fun onDestroy() {
    mDoubleBackHandler.removeCallbacks(mExitRunnable)
    super.onDestroy()
  }

  // endregion

  // region Helpers

  private fun setTabIcon(tab: TabLayout.Tab?, @DrawableRes icon: Int) {
    if (tab == null) {
      return
    }

    tab.setIcon(icon)
  }

  private fun setTabColor(tab: TabLayout.Tab?, color: Int) {
    tab?.icon?.setColorFilter(color, PorterDuff.Mode.SRC_IN)
  }

  private fun setLayout() {
    setTheme(R.style.AppTheme)
    setContentView(R.layout.main_activity)
  }

  // endregion
}
